proposals/LP0004-project-governance.md

```diff
@@ -0,0 +1,393 @@
+
+Second, _area teams_ are responsible for maintaining an up-to-date and comprehensive list of code owners for their area of the project. They can nominate any individual they deem appropriate as code owner of any area they are responsible for. The role of _code owner_ remains a volunteer role, and any individual can accept, decline, or resign the role for themselves as they feel appropriate.
+
+Finally, _area teams_ are responsible for facilitating decision making for their area of the project. Facilitating decision making can take any number of forms ranging from contributing to RFC discussions, helping mediate disagreements, or fulfilling roles originally delegated to Chris Lattner in the [LLVM Decision Making](https://github.com/llvm/llvm-www/blob/main/proposals/LP0001-LLVMDecisionMaking.md) process.
+
+When acting to facilitate decision making the _area team _should act as a mediator between different perspectives helping find common ground and recognizing that decisions need not be binary. The _area team_ should seek to find the best solution to the framed problem, which may not be any of the proposed alternatives. If agreement cannot be reached, the _area team_ may act as the final decision maker. In that capacity decisions of an _area team_ are considered final, but can be overruled by a 2/3 majority vote of the _project council_ or the _area team_ itself revisiting the issue.
+
+_Area teams_ are not intended to be direction setters or primary maintainers of their areas, although individuals on an _area team_ may fulfill that role separately. The _area team’s_ role is as a steward and moderator ensuring the health and smooth operation of the area.
+
+Each _area team_ will have 5 members elected by the _voting contributors_. Candidates for _area teams_ must be a _voting contributor_ and self-nominated. An individual cannot serve on two _area teams_.
+
+Members of an _area team_ are elected for 1 year terms. Individuals are encouraged to not run for more than two consecutive terms. To encourage turnover, incumbent members of an _area team_ must place in the top 4 candidates during an election. The fifth spot will always go to a non-incumbent candidate.
+
+
+#### Initial Area Teams
+
+With this proposal a group of initial _area teams_ will be formed with defined domain areas. The members of each _area team_ will be defined by an election held after the adoption of the proposal. The initial _area teams_ proposed below are based on statistics generated from the LLVM project repository (See Appendix 2 for applicable scripts).
+
+In dividing the LLVM project into _area teams_, this proposal seeks to break up areas of the project based on contribution volume and number of active contributors. This proposal assumes that areas of the project with more unique contributors are more likely to have contributor disagreements, and that areas with more highly active contributors will be more likely to have diverse _area team_ formations.
```

Comment:
Dividing the project up into area teams is probably one of the most difficult parts of this proposal. Appendix 1 describes the metrics methodology I used (I'm going to add a big bolded note about how insufficient that data is, because it is insufficient).

One of my goals in splitting into area teams was to construct areas that had enough contributors that were active enough to produce area teams that cycled membership. I also looked to try and group areas of the project that had similarities. Flang is an obvious outlier where I just didn't quite know where to fit it.

I looked at a couple different bits of data, but the key metrics were: number of commits, number of authors, and commits per author. On that last point I was particularly interested in how quickly the number of commits per author fell off. For example, if you look at `lld` and `lldb` which each had similar numbers of total authors, they had very different fall off curves:

```
./llvm-project/lld (130 authors):
  Contributor 1: 155
  Contributor 2: 54
  Contributor 3: 29
  Contributor 4: 26
  Contributor 5: 22
  Contributor 6: 22
  Contributor 7: 17
  Contributor 8: 16
  Contributor 9: 12
  Contributor A: 11
./llvm-project/lldb (149 authors):
  Contributor 1: 174
  Contributor 2: 159
  Contributor 3: 141
  Contributor 4: 107
  Contributor 5: 101
  Contributor 6: 100
  Contributor 7: 98
  Contributor 8: 58
  Contributor 9: 52
  Contributor A: 43
```

My bias here was that `LLDB` would be likely to have more contributors who could cycle through a leadership role than `LLD`. This is in part because I think people who only contribute 0-2 times per month are likely less interested in being involved in project governance. I have no evidence to justify that belief other than a strong suspicion.

Reply:
> This is in part because I think people who only contribute 0-2 times per month are likely less interested in being involved in project governance.

I don't think there's a direct correlation there. It may be reasonable to say "the _probability_ of finding people interested is higher" but just because there's more people, not because a person who commits often is interested in _project governance_.

There's also many artifacts that can generate high commit rate (ex. bot CI reverts, repeated fixups, people that don't squash their PRs), which makes it a really bad metric. 

Even number of PRs merged is bad. On a small sub-community, two or three developers can approve each other's patches (with good intentions and for the right reasons) much quicker than a large refactor in Clang.

Lines of code is also another bad one. I can easily beat that (unintentionally) by renaming variables or updating some tests, or adding a new benchmark.

There is no good metric that cannot be (even unintentionally) broken or abused. The best approach is to not even begin to rely on metrics.

How about we ask the people in the sub-communities how they want to be divided? @AaronBallman did an awesome job reorganizing Clang's code ownership. I would personally not have done that way (too complex), but the rest of the Clang folks seemed really happy with it, so it made me happy, too.